src/Analyser/ArgumentsNormalizer.php

```diff
@@ -23,12 +23,60 @@ final class ArgumentsNormalizer
 
 	public const ORIGINAL_ARG_ATTRIBUTE = 'originalArg';
 
+	public static function reorderCallUserFuncArguments(
+		FuncCall $callUserFuncCall,
+		Scope $scope,
+	): ?FuncCall
+	{
+		$args = $callUserFuncCall->getArgs();
+		if (count($args) < 1) {
+			return null;
+		}
+
+		$calledOnType = $scope->getType($args[0]->value);
```

Comment:
Still wrong. See: https://3v4l.org/F67Rj

Reply:
added a test, which shows it works If I got your point.

at this point of execution named args are already re-ordered to positional ones

Reply:
Your logic isn't correct because I don't see the handling of `call_user_func` that would be necessary for PHPStan to understand it.

I added a rule to prove my point. You can see the rule doesn't see the signature correctly and so can't the dynamic return type extension.

The problem is that MutatingScope (before it calls your extension) already calls ArgumentsNormalizer which mangles the arguments so they're wrong. Because it doesn't know what to do with them.

Reply:
Thank you. I have an idea now what you are talking about.

Do you expect this PR to still contain this new rule when finished, or do you think I should/can adjust the core in a way, that the rewritten FuncCall will be finally checked by the pre-existing `CallToFunctionParametersRule` ?

Reply:
I've got an idea:

1) It should not be an extension, just a piece of code in MutatingScope before extensions are called. Before the normalizer is used.
2) That way we can find the first/callback arg and remove it from args. And create a new FuncCall with the rest.
2) We can call $this->getType on the FuncCall and return that.
4) The new rule should just do the same thing as in 2) and pass the new FuncCall to FunctionCallParametersCheck.

Reply:
thanks. 

I had to re-calibrate my expectations for this PR, as it turned from a return-type improving into a more overall goal.
it seems to work as you described now.